### PR TITLE
#855 Improve test coverage of lensing module

### PR DESCRIPTION
**Pull request by rmjarvis**
This PR is mostly just additional tests to improve the test coverage of lensing_ps.py, nfw_halo.py and pse.py.  In a couple cases I marked stuff as `pragma: no cover` rather than write the tests for functions that are never really used within GalSim.  And in a couple other cases where we never use options, I went ahead and deprecated them.

Deprecated:
* passing `Image` arguments to `kappaKaiserSquires` function
* `interpolant` argument for PowerSpectrum methods `getShear`, `getConvergence`, `getMagnification`, and `getLensing`.  

Explicitly uncovered:
* `calculateXi` (There is some coverage, but not nearly for all the possible parameter options, and I didn't want to bother with it.  This was relevant during Great3, but I'm not sure how useful it is anymore.  I wouldn't mind deprecating this function.)
* `subsampleGrid`  (Never used by anything.  Again, I wouldn't mind deprecating it.)

API change:
* The return values from `getShear`, etc. for both `NFWHalo` and `PowerSpectrum` used to try to match the input type being a tuple, list, or numpy array.  Now if it gets any kind of list-like input, it just returns a numpy array.  Python duck typing should suffice for essentially all uses here, so I think this will be transparent to all users, but technically it is an API change.

The other item that is now covered, but might nonetheless merit deprecation is the `weight_*` options for `PowerSpectrumEstimator`.  We never use them (or `theory_func`), and when I wrote unit tests for them, the accuracy is significantly worse than not weighting.  If someone can make a case for why they are useful, we should keep them, but if not, I'm inclined to deprecate them.  @joezuntz may have something to say on this, since he was the original author of pse.py.

BTW, I actually started this over a year ago, then never got back to it, and ended up forgetting I had even started it.  But since @gijzelaerr ran into a bug in one of our uncovered error messages, I figured it was worth doing, so I finished this up.  This PR would resolve both #855 and #945.

### CODE REVIEW

File: galsim/lensing_ps.py [hunk 1]
```diff
@@ -33,43 +33,17 @@ def theoryToObserved(gamma1, gamma2, kappa):
...
-        return float(g1), float(g2), float(mu)
-    elif isinstance(gamma1, list):
-        return list(g1), list(g2), list(mu)
-    elif isinstance(gamma1, tuple):
-        return tuple(g1), tuple(g2), tuple(mu)
-    elif isinstance(gamma1, np.ndarray):
-        return g1, g2, mu
-    else:
-        raise ValueError("Unknown input type for shears, convergences: %s",type(gamma1))
+    g1 = gamma1/(1.-kappa)
```
Comment (jmeyers314) [inline galsim/lensing_ps.py]: Maybe leave open the possibility of passing in a python list (similar to LookupTable e.g.)?  I.e.
```python
gamma1 = np.array(gamma1, dtype=float)
```
Simplifying the output is definitely a good idea.

File: galsim/nfw_halo.py [hunk 2]
```diff
@@ -195,91 +184,66 @@ def __omega(self, a):
...
-        if out is None:
-            out = np.zeros_like(x)
+        out = np.zeros_like(x, dtype=float)
 
         # 3 cases: x > 1, x < 1, and |x-1| < 0.001
-        mask = (x < 0.999)
-        if mask.any():
-            a = ((1.-x[mask])/(x[mask]+1.))**0.5
-            out[mask] = 0.5*np.log((1.+a)/(1.-a))/(1-x[mask]**2)**0.5
+        mask = np.where(x < 0.999)[0]
```
Comment (jmeyers314) [inline galsim/nfw_halo.py]: Why `np.where` (Here and below)?  I think the original is easier to read.  (Don't have to remember why there's a `[0]`).  Getting rid of the `mask.any()` is a definite improvement though.

File: tests/test_lensing.py [hunk 3]
```diff
@@ -617,6 +707,18 @@ def test_shear_units():
...
+    ps = galsim.PowerSpectrum('k', 'k**2', False, 'arcsec')
+    assert ps == galsim.PowerSpectrum(e_power_function='k', b_power_function='k**2')
+    assert ps == galsim.PowerSpectrum(e_power_function='k', b_power_function='k**2',
+                                      delta2=False, units=galsim.arcsec)
+    for ps2 in [ galsim.PowerSpectrum('k**2', 'k**2', False, 'arcsec'),
+                 galsim.PowerSpectrum('k', 'k', False, 'arcsec'),
+                 galsim.PowerSpectrum('k', 'k**2', True, 'arcsec'),
+                 galsim.PowerSpectrum('k', 'k**2', False, 'arcmin'),
+               ]:
+        assert ps != ps2
```
Comment (jmeyers314) [inline tests/test_lensing.py]: `all_obj_diff` (here and for `PowerSpectrumRealizer`)?

File: galsim/nfw_halo.py (hunk 2 above)
Comment (rmjarvis) [reply galsim/nfw_halo.py]: The reason for this is efficiency.  Indexing with an array of bools tends to be slower than indexing with an array of integers saying where the bools were True.  Especially if there are far fewer True than False, which I suppose will be the case for typically 2 out of these 3 cases.

**Comment by rmandelb** [comment]
Hi @rmjarvis - I would indeed like to take a look.  Will do so this weekend.  Thanks!

**Comment by rmandelb** [comment]
Before I look at the code, a few comments:

- I feel like passing different interpolants to `getLensing` (etc.) is a useful thing to be able to do, at least to be able to check that you aren't that sensitive to the choice of interpolant (because if you are, that usually means you've set up the grid in a way that is suboptimal for the problem of choice).

- I agree about deprecating `subsampleGrid` (which you haven't done yet but I thought I would comment since you mentioned it as an option).  That was mostly for testing when we were putting the basic functionality in place.

- I personally do use `calculateXi` still.  It's useful to be able to predict how much suppression of shear power you'll have due to your finite grid size.  We could deprecate some arguments while keeping the basic functionality?

**Comment by rmandelb** [comment]
I looked at the code and didn't really have anything to add - the new tests and other changes look good.  I edited my above comment about planned changes, so please read on GitHub rather than by e-mail.

**Comment by rmjarvis** [comment]
Thanks Rachel.  I deprecated subsampleGrid.  

And since you say you use calculateXi, I went ahead and added coverage for it.  Technically, I don't test all the possible input parameters, but I refactored the code slightly, so it now shares the implementation for handling delta2, scale, and bandlimit with buildGrid.  So I think that's probably sufficient coverage.

Also, there were indeed bugs with how calculateXi handled B-modes, which the tests uncovered, and which I fixed.

**Comment by rmandelb** [comment]
Thanks Mike.  Indeed our previous calculateXi tests were done only with E-mode power spectra.

**Pull request merged**